### ISSUE TITLE
Add More Shaders

### ISSUE THREAD

**Issue by wingedonezero** [mgba-emu/mgba#304]
I haven't been able to find any information on the adding of shaders or compatible ones. Can you add a hq4x and just a normal 5xbrc version? I dont quite understand the options for the xbr shader included i would just like a simple version that doesn't need configured unless you have a guide to understand what the options do? But to me hq4x is pretty standard in most emulators and 5xbrc's rounding just looks nice on most games. Any insight would be appriciated.

**Comment by endrift** [comment]
You don't really need to touch the options at all. I barely understand what they do; I just ported one of Hyllian's xBR shaders to GLSL. However, this one's really buggy so I might replace it at some point. hq4x is rather obsolete at this point, but if it's not too much trouble I might add it. I've never even heard of 5xbrc.

**Comment by wingedonezero** [comment]
Well its 5xbr version c and the shader does work but for some reason lvl2 just feels like it blurs a bit more to me. If i can ask why didnt you just use .glsl shaders instead of the vs and fs theres a ton of them and you wouldnt have to port each and everyone. People could just use the ones from retro arch or you could include a select few i tryed to make one of the mainfest and use it by copying the info from the xbr one and adding the filename but that didnt quite work out. Unless you have some tips to try to make it work.

**Comment by ImSpecial** [comment]
I came to request this, and I see it has already been done, so I'm just going to add my voice to this to maybe show there is some demand for this request. 

VBA-M has more shader options and I'd love to just get rid of it altogether but some of the shaders are really nice and I keep it around for this, SuperEagle for example looks great with some games. I don't really know much about making shaders myself nor fine tuning the files to make 'em look just right either, but there are tons of shader packs out there for SNES9X, Gens/Fusion, Nestopia, ePSXe, etc. and if some of these could just be copied over into a 'shader' folder in mGBA's directory, well that would be lovely.

**Comment by inukaze** [comment]
Really the mgba needs more shaders. 

Because the unique i have is "xbr" (But is level 3) i need max level 2 , because is extreme slowly on my pc . well someone have a manual how to port RetroArch Shaders CG / GLSL , to mgba-qt ??? 

i wanna try : xsai , x2sai , xbr (level 1, level 2) , 2xsai , SuperEagle, LQ2X , HQ2X , Bicubic, , SRGB and Retro with mgba-qt

**Comment by endrift** [comment]
Get people to write them, and I'll merge them. Otherwise, this is about the lowest priority I have.

**Comment by inukaze** [comment]
Precisely for that, ask if you do not have a guide / manual, how to manually convert the shaders available in retroarch.

It's to do them on my own, since I'm not a programmer, but I still want more shaders for mgba

**Comment by endrift** [comment]
The RetroArch shaders are in Cg format, which is a pain to convert to the GLSL format that mGBA uses.

That said, I've added xBR level 2.

**Comment by inukaze** [comment]
Where I can download it?

**Comment by endrift** [comment]
Either download the next nightly (which will be available in about 5 hours) or download the files in this folder: https://github.com/mgba-emu/mgba/tree/master/res/shaders/xbr-lv2.shader

**Comment by dojima** [comment]
In the RetroArch package, there's a 'shaders_glsl' folder that has a bunch of files in *.glslp/*.glsl formats.  Are those in the correct format to use with mGBA?  I dunno the etiquette behind software authorship, but would it be all right to use them for mGBA if so?  I tried making use of them myself, but I'm guessing it's more involved than simply moving the files over and renaming things, haha.

**Comment by Tharn** [comment]
I don't think plain .glsl shaders are readable. You'll notice that mGBA includes a .manifest file which has some settings like using different .fs files for different shader passes, as well as color settings for mGBA itself.

I would also add my vote for more shaders. There are plenty of LCD-alikes and color shaders already in mGBA, but not many real upscalers to choose from. HQ shaders would be nice. As well as EPX (my favourite from Snes9x), BRZ, Eagle, SaI.

**Comment by dojima** [comment]
Yeah, I resorted to just using the mGBA core within RetroArch.  I find RetroArch to be overly contrived and unintuitive, but those shaders make its use somewhat of a necessary evil for me.

**Comment by Tharn** [comment]
I tried, at least. (#1297)

**Comment by Senhordelico** [comment]
Hello. First of all, I would like to say this: mGBA is great, but I think is missing a little one thing. Why don't you allow the emulador to mix two shaders? For example: I love the retro-style of scanlines and ctr shaders, but it will great to allow one more shaders, like scanlines and xBRz together, for example. The player could mix the shaders, bringing more options to alter the game image.  I don't know if is possible, but will be good increase for the emulator.

**Comment by endrift** [comment]
"Mixing" shaders is covered by #864

**Comment by Anuskuss** [comment]
@endrift What are your thoughts on GBA "borders" like:
![https://forums.libretro.com/t/handheld-border-shaders/2551](https://i.postimg.cc/bwKzRwxr/Sonic-Advance-Europe-En-Ja-Fr-De-Es-170719-1.png)

I play in fullscreen but GBA games are not really meant for big TVs so shrinking the screen down (more than integer scaling can do) would make playing more enjoyable to me.

**Comment by endrift** [comment]
mGBA shaders cannot currently handle textures (so no images). That said, I am planning to add custom borders soon.

**Comment by Jules-A** [comment]
Probably not the best issue to ask this on (I can't find the original post) but are you still planning on allowing external shaders? 

By that I think you originally meant shaders that wouldn't be included in the main mGBA download but could be downloaded separately and placed in a shader folder to get around licensing issues.

**Comment by endrift** [comment]
I've ported a handful of more shaders from SameBoy, including its hq2x (it doesn't have 3x or 4x) and OmniScale, which seems to be like hqx on crack in some regards.

**Comment by v-fox** [comment]
> I've ported a handful of more shaders from SameBoy, including its hq2x (it doesn't have 3x or 4x) and OmniScale, which seems to be like hqx on crack in some regards.

Nice one.
But, unfortunately, that "omni" scaler doesn't seem to be "resolution-independent" as it claims. As with all others, with 2x it loses almost all effect and anything above is pure integer scaling. There is also some kind of line artefacts.

Tested both with Aria of Sorrow ([Recolor](https://www.romhacking.net/hacks/6301/) hack) and for me the best result (some rounding without being too aggressive and losing details) is still seems to be: 2x resolution + xbr-lv3 with 2/24/21/4 settings. But hq2x with 1x resolution seems decent.

**Comment by crtshaders** [comment]
Hi endrift! 

I want to learn how to port shaders, I'm not good at programming but if I have a good source material and tutorials I'm sure that can do it. 

I'm having audio crackling with retroarch core so that's the reason why I'm using standalone version but doesn't have crt shaders. I want to port this one [https://github.com/libretro/glsl-shaders/blob/master/crt/shaders/zfast_crt_nogeo_svideo.glsl](https://github.com/mgba-emu/mgba/issues/url)

The problem is that i don't understand mgba shaders strcture like vs,fs and ini so I don't know how to port the shader.

Greetings.

**Comment by thekryz** [comment]
I just stumbled across this project and I love it!
That said, if libretro shaders are possible, I would love to see the [Scalefx-Hybrid](https://github.com/libretro/common-shaders/blob/master/scalefx/scalefx-hybrid.cgp) shader (ScaleFX + Reverse AA) implemented.

**Comment by charleshuang3** [comment]
I just saw the mmpx shader, seems very interesting https://casual-effects.com/research/McGuire2021PixelArt/index.html

Then I convert this [reshare 4 shader impl](https://gist.github.com/martymcmodding/ed60af0f8b928c59ce2cabd5292ff11a) to glsl 

https://gist.github.com/charleshuang3/3f08ded9cd6b665b0195d741986e8084. I load this to mgba, seems work (at least not black screen). I am not familiar with sharder. feel free to make it formal and add to mgba.

**Comment by v-fox** [comment]
> https://gist.github.com/charleshuang3/3f08ded9cd6b665b0195d741986e8084. I load this to mgba, seems work (at least not black screen). I am not familiar with sharder. feel free to make it formal and add to mgba.

Didn't work for me, so I've tried porting [version from libretro](https://github.com/libretro/glsl-shaders/blob/468f67b6f6788e2719d1dd28dfb2c9b7c3db3cc7/scalenx/shaders/mmpx.glsl). Miraculously, it did work: [MMPX.shader.zip](https://github.com/user-attachments/files/24282480/MMPX.shader.zip)
I've decided to use bundled `/usr/share/mgba/shaders/scale4x.shader` for now due to better smoothness/detail ratio.  But maybe MMPX just needs some tunning.

While trying to get rid of shader compilation errors, I've built mgba with:
```
	sed -e 's:version 300 es:version 320 es:g' \
		-e 's:version 330 core:version 430 core:g' \
		-i src/gba/renderers/gl.c
	sed -e 's:#version 150 core:#version 430 core:g' \
		-e 's:format.setVersion(3, 3):format.setVersion(4, 3):' \
		-e 's:format.setVersion(3, 1):format.setVersion(3, 2):' \
		-i src/platform/qt/DisplayGL.cpp
```
Also `-DBUILD_GL=ON -DBUILD_GLES2=ON -DBUILD_GLES3=ON` and SDL3 imitating SDL2.

**Comment by endrift** [comment]
I've done a simple port of MMPX in 583eb8eb98d0000bcd1260050580a043a02c4170 based on [the original paper's implementation](https://casual-effects.com/research/McGuire2021PixelArt/index.html)
